### PR TITLE
Allow instantiation of a Trie using Poseidon hash

### PR DESCRIPTION
**Pull request by omerfirmak**

### CODE REVIEW

File: core/trie/trie.go [hunk 1]
```diff
@@ -38,14 +38,26 @@ type Trie struct {
...
 }
 
+type NewTrieFunc func(Storage, uint, *bitset.BitSet) *Trie
+
 func NewTrie(storage Storage, height uint, rootKey *bitset.BitSet) *Trie {
+	return newTrie(storage, height, rootKey, crypto.Pedersen)
+}
+
+func NewTriePoseidon(storage Storage, height uint, rootKey *bitset.BitSet) *Trie {
+	return newTrie(storage, height, rootKey, crypto.Poseidon)
```
Comment (omerfirmak) [inline core/trie/trie.go:51]: Uh, I really dont want to write tests for this just to be able to merge.

**Comment by omerfirmak** [comment]
I will include this in to a bigger PR.

File: core/trie/node_test.go [hunk 2]
```diff
@@ -170,5 +171,5 @@ func TestNodeHash(t *testing.T) {
 	}
 	path := bitset.FromWithLength(6, []uint64{42})
 
-	assert.Equal(t, true, expected.Equal(node.Hash(path)), "TestTrieNode_Hash failed")
+	assert.Equal(t, true, expected.Equal(node.Hash(path, crypto.Pedersen)), "TestTrieNode_Hash failed")
```
Comment (IronGauntlets) [inline core/trie/node_test.go:174]: Please add a case for Poseidon

File: core/trie/trie.go [hunk 3]
```diff
@@ -38,14 +38,26 @@ type Trie struct {
 	height  uint
 	rootKey *bitset.BitSet
 	storage Storage
+	hash    func(*felt.Felt, *felt.Felt) *felt.Felt
 }
 
+type NewTrieFunc func(Storage, uint, *bitset.BitSet) *Trie
+
 func NewTrie(storage Storage, height uint, rootKey *bitset.BitSet) *Trie {
```
Comment (IronGauntlets) [inline core/trie/trie.go:46]: Please rename `NewTrie` to `NewTriePedersen`

File: core/trie/trie.go [hunk 4]
```diff
@@ -38,14 +38,26 @@ type Trie struct {
 	height  uint
 	rootKey *bitset.BitSet
 	storage Storage
+	hash    func(*felt.Felt, *felt.Felt) *felt.Felt
 }
 
+type NewTrieFunc func(Storage, uint, *bitset.BitSet) *Trie
```
Comment (IronGauntlets) [inline core/trie/trie.go:44]: I don't see any usage of this. If this is not being used please remove

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: core/trie/node_test.go (hunk 2 above)
Comment (omerfirmak) [reply core/trie/node_test.go:174]: That would be testing the Poseidon hash's implementation, not TrieNode.Hash. Testing for each dependency (hash function in this case) defeats the purpose of dependency injection.  

and we don't have a test vector for Poseidon tries.

File: core/trie/trie.go (hunk 3 above)
Comment (omerfirmak) [reply core/trie/trie.go]: done

File: core/trie/trie.go (hunk 4 above)
Comment (omerfirmak) [reply core/trie/trie.go]: done

File: core/trie/trie.go [hunk 5]
```diff
@@ -38,14 +38,26 @@ type Trie struct {
 	height  uint
 	rootKey *bitset.BitSet
 	storage Storage
+	hash    func(*felt.Felt, *felt.Felt) *felt.Felt
```
[the remark this answers is not in the document]
Comment (omerfirmak) [reply core/trie/trie.go]: done

File: core/trie/trie.go [hunk 6]
```diff
@@ -17,6 +17,8 @@ type Storage interface {
 	Delete(key *bitset.BitSet) error
 }
 
+type HashFunc func(*felt.Felt, *felt.Felt) *felt.Felt
+
```
Comment (IronGauntlets) [inline core/trie/trie.go:21]: Any particular reason this is public? 

We should only make it public if `NewTrie` is also public. Which is currently not the case.

**Review by IronGauntlets — CHANGES_REQUESTED** [review]

File: core/trie/trie.go (hunk 6 above)
Comment (omerfirmak) [reply core/trie/trie.go]: made private

**Review by IronGauntlets — APPROVED** [review]

**Pull request merged**